File: 상남/구현/혼자 놀기의 달인.py
def solution(cards):
    n = len(cards)
    visited = [False] * (n + 1)
    rst = []
    # for i in range(1,n+1):
    #     data = []
    #     if visited[i] == False:
    #         # visited[i] = True
    #         data.append(i)
    #         tmp = i
    #         while True:
    #             if visited[tmp] == False:
    #                 tmp = data[tmp-1]
    #                 data.append(tmp)
    #                 visited[tmp] = True
    #             else:
    #                 rst.append(data)
    #                 break
    # while cards:
    rst = []
    for i in range(1, n + 1):
        data = []
        if visited[i] == False:
            data.append(i)
            visited[i] = True
            tmp = i

            while True:
                tmp = cards[tmp - 1]

                if visited[tmp] == False:
                    data.append(tmp)
                    visited[tmp] = True
                else:
                    rst.append(data)
                    break
    answer = []
    for i in rst:
        answer.append(len(i))
    # print(answer)
    answer.sort(reverse=True)
    if answer[0] == n: # 이 예외문을 생각치 못했다;;
        return 0
    else:
        return answer[0] * answer[1]

File: 현준/24-1/solve_24_04_week_02/혼자_놀기의_달인.java
import java.util.ArrayList;
import java.util.Arrays;

class 혼자_놀기의_달인 {

    // 24.04.09 (화)

    public int solution(int[] cards) {
        boolean[] opened = new boolean[cards.length + 1]; // 상자 오픈 여부
        int[][] map = new int[cards.length + 1][1]; // 각 카드가 가리키는 상자는 하나만 존재, 그렇기에 2차원 배열에 내부 배열 사이즈를 1로 설정
        for (int i = 1; i <= cards.length; i++) { // "cards[i]는 i + 1번 상자에 담긴 카드에 적힌 숫자를 의미합니다."
            map[i][0] = cards[i - 1]; // 1부터 시작하는 것으로 조정했기 때문에 -1
        }

        /*
        for (int i = 1; i <= cards.length; i++) {
            System.out.println(i + " " + map[i][0]);
        }
        */

        ArrayList<Integer> result = new ArrayList<>();
        for (int i = 1; i <= cards.length; i++) {
            if (!opened[i]) { // 열리지 않았을 때만 그룹에 추가
                result.add(logic(i, map, opened));
            }
        }

        if (result.size() == 1) { // "만약 1번 상자 그룹을 제외하고 남는 상자가 없으면 그대로 게임이 종료되며, 이때 획득하는 점수는 0점입니다."
            return 0;
        }

        int[] answer = result.stream().mapToInt(x -> x).toArray();
        Arrays.sort(answer);

        // "1번 상자 그룹에 속한 상자의 수와 2번 상자 그룹에 속한 상자의 수를 곱한 값이 게임의 점수입니다."
        // "범희가 이 게임에서 얻을 수 있는 최고 점수를 구해서 return 하도록 solution 함수를 완성해주세요."
        return answer[answer.length - 1] * answer[answer.length - 2];
    }

    // 그룹의 갯수 계산 함수
    private int logic(int index, int[][] map, boolean[] opened) {
        int sum = 0;
        int start = index;
        while (!opened[start]) { // 열리지 않았을 때만 계속 진행
            opened[start] = true;
            start = map[start][0]; // ex: 1번 상자에 3이 적혀있으면 3번 상자로 이동
            sum++;
        }
        return sum;
    }
}
